Inject `compiler_builtins` during postprocessing rather than via AST

tgross35 opened the pull request:
`compiler_builtins` is currently injected into the AST as `extern crate compiler_builtins as _`. This has made gating via diagnostics difficult because it appears in the crate graph as a non-private dependency, and there isn't an easy way to differentiate between the injected AST and user-specified `extern crate compiler_builtins`.

Resolve this by injecting `compiler_builtins` during postprocessing rather than early in the AST. Most of the time this isn't even needed because it shows up in `std` or `core`'s crate graph, but injection is still needed to ensure `#![no_core]` works correctly.

A similar change was attempted at [1] but this encountered errors building `proc_macro` and `rustc-std-workspace-std`. Similar failures showed up while working on this patch, which were traced back to `compiler_builtins` showing up in the graph twice (once via dependency and once via injection). This is resolved by not injecting if a `#![compiler_builtins]` crate already exists.

This PR is a smaller subset of changes, split from https://github.com/rust-lang/rust/pull/135501. #135501 builds on this to resolve a couple remaining privacy issues and actually do the update that has been troublesome.

[1]: https://github.com/rust-lang/rust/pull/113634

try-job: aarch64-gnu
try-job: dist-x86_64-linux
try-job: i686-mingw
try-job: x86_64-gnu-distcheck
try-job: x86_64-msvc-1
try-job: x86_64-rust-for-linux
try-job: test-various

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13024492203/job/36331216841) [(plain)](https://github.com/rust-lang/rust/commit/00390e8b77c326f5dfcee5c4c592719178f4767a/checks/36331216841/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#22 exporting to docker image format
#22 sending tarball 30.1s done
#22 DONE 35.7s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [ui] tests/ui/attributes/log-backtrace.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 101
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUSTC_LOG="info" RUSTC_LOG_BACKTRACE="rustc_metadata::creader" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/attributes/log-backtrace.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "-O" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/attributes/log-backtrace/a" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
--- stderr -------------------------------
<<<<<< TRUNCATED, SHOWING THE FIRST 524562 BYTES >>>>>>

 INFO rustc_metadata::creader resolving crate `std`
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13024940512/job/36332363160) [(plain)](https://github.com/rust-lang/rust/commit/4bd1a8c3ff02a5557685fc997bf74bc2ac7fe420/checks/36332363160/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 28.1s done
#21 DONE 31.1s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
   Compiling rustc_monomorphize v0.0.0 (/checkout/compiler/rustc_monomorphize)
   Compiling rustc_ast_lowering v0.0.0 (/checkout/compiler/rustc_ast_lowering)
   Compiling rustc_query_impl v0.0.0 (/checkout/compiler/rustc_query_impl)
   Compiling rustc_smir v0.0.0 (/checkout/compiler/rustc_smir)
error[E0559]: variant `CrateOrigin<'_>::AstExtern` has no field named `name`
    --> compiler/rustc_metadata/src/creader.rs:1208:25
1208 |                         name,
1208 |                         name,
     |                         ^^^^ `CrateOrigin<'_>::AstExtern` does not have this field
     = note: all struct fields are already assigned


error[E0559]: variant `CrateOrigin<'_>::AstExtern` has no field named `orig_name`
    --> compiler/rustc_metadata/src/creader.rs:1209:25
1209 |                         orig_name,
1209 |                         orig_name,
     |                         ^^^^^^^^^ `CrateOrigin<'_>::AstExtern` does not have this field
     = note: all struct fields are already assigned

error: unused import: `CrateStore`
  --> compiler/rustc_metadata/src/creader.rs:28:32
  --> compiler/rustc_metadata/src/creader.rs:28:32
   |
28 |     CrateDepKind, CrateSource, CrateStore, ExternCrate, ExternCrateSource,
   |
   = note: `-D unused-imports` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(unused_imports)]`

```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13025058963/job/36332675607) [(plain)](https://github.com/rust-lang/rust/commit/b895c5c9aa9a7122ed954171852d6db1d5570a11/checks/36332675607/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#22 exporting to docker image format
#22 sending tarball 28.1s done
#22 DONE 34.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
+ error: `#[panic_handler]` function required, but not found
+ 
+ error: unwinding panics are not supported without std
+    |
+    = help: using nightly cargo, use -Zbuild-std with panic="abort" to avoid unwinding
+    = note: since the core library is usually precompiled with panic="unwind", rebuilding your crate with panic="abort" may not be enough to fix the problem
1 error: requires `sized` lang_item
2 
- error: aborting due to 1 previous error
+ error: aborting due to 3 previous errors
---
To only update this specific test, also pass `--test-args proc-macro/issue-59191-replace-root-with-fn.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/proc-macro/issue-59191-replace-root-with-fn.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/issue-59191-replace-root-with-fn" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/issue-59191-replace-root-with-fn/auxiliary" "--edition=2018" "--extern" "issue_59191=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/issue-59191-replace-root-with-fn/auxiliary/libissue_59191.so"
--- stderr -------------------------------
error: `#[panic_handler]` function required, but not found

error: unwinding panics are not supported without std
error: unwinding panics are not supported without std
   |
   = help: using nightly cargo, use -Zbuild-std with panic="abort" to avoid unwinding
   = note: since the core library is usually precompiled with panic="unwind", rebuilding your crate with panic="abort" may not be enough to fix the problem
error: requires `sized` lang_item

error: aborting due to 3 previous errors
------------------------------------------
```

</details>

tgross35 committed:
Track where crate dependencies come from in `creader`

Introduce an enum that represents the different possible sources for
dependencies, and use them where possible. This will enable more fine
grained control and provides better context than passing the `dep_root`
tuple.

Use this to ensure that injected crates always show up as private by
default.

tgross35 commented:
The current bug is that `compiler_builtins` depends on `rustc_std_workspace_core`, which for some reason is resolving its dependency `core` with a different `cnum` than the first time `core` was loaded. I think this is the same behavior @amanieu was running into at https://github.com/rust-lang/rust/pull/113634#issuecomment-1635043738.

```text
 INFO rustc_metadata::creader POSTPROCESS
 INFO rustc_metadata::creader resolving crate `compiler_builtins`
 INFO rustc_metadata::creader falling back to a load
 INFO rustc_metadata::creader register newly loaded library for `compiler_builtins`
 INFO rustc_metadata::creader register crate `compiler_builtins` (cnum = 14. private_dep = true)
 INFO rustc_metadata::creader resolving dep `compiler_builtins`->`rustc_std_workspace_core` hash: `f67002fa0554298ad26d2bf8107374d1` extra filename: `-0f3268257eb1faf9`
 INFO rustc_metadata::creader resolving crate `rustc_std_workspace_core`
 INFO rustc_metadata::creader falling back to a load
 INFO rustc_metadata::creader register newly loaded library for `rustc_std_workspace_core`
 INFO rustc_metadata::creader register crate `rustc_std_workspace_core` (cnum = 15. private_dep = false)
 INFO rustc_metadata::creader resolving dep `rustc_std_workspace_core`->`core` hash: `fc48f0a1c1d99c8755223ff791e4cabc` extra filename: `-af66f1d2fa6253c2`
 INFO rustc_metadata::creader resolving crate `core`
 INFO rustc_metadata::creader falling back to a load
 INFO rustc_metadata::creader register newly loaded library for `core`
 INFO rustc_metadata::creader returning false for core is private
 INFO rustc_metadata::creader register crate `core` (cnum = 16. private_dep = false)
 INFO rustc_metadata::creader resolving dep `compiler_builtins`->`core` hash: `fc48f0a1c1d99c8755223ff791e4cabc` extra filename: `-af66f1d2fa6253c2`
 INFO rustc_metadata::creader resolving crate `core`
 INFO rustc_metadata::creader library for `core` was loaded previously, cnum 16
 INFO rustc_metadata::creader returning false for core is private
 INFO rustc_metadata::creader panic runtime injection skipped, only generating rlib
``

tgross35 commented:
Actually it looks like `compiler_builtins` is getting loaded twice, once as a dependency of `std` and once via injection, which gives it two different hashes. Going to try skipping injection if it is already in the graph.

tgross35 commented:
@bors try

bors commented:
:hourglass: Trying commit 2a7db5646efcc22695554d84b9da3717c4b52445 with merge 05b62dfcb12c144f47af05973134b99169a808b2...
<!-- homu: {"type":"TryBuildStarted","head_sha":"2a7db5646efcc22695554d84b9da3717c4b52445","merge_sha":"05b62dfcb12c144f47af05973134b99169a808b2"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13042572106/job/36387385393) [(plain)](https://github.com/rust-lang-ci/rust/commit/05b62dfcb12c144f47af05973134b99169a808b2/checks/36387385393/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Inject `compiler_builtins` directly rather than in AST
test-various
x86_64-msvc
##[endgroup]
INFO:root:Job type: TryRunType(custom_jobs=['test-various', 'x86_64-msvc'])
  File "/home/runner/work/rust/rust/src/ci/github-actions/ci.py", line 314, in <module>
    calculate_job_matrix(data)
  File "/home/runner/work/rust/rust/src/ci/github-actions/ci.py", line 266, in calculate_job_matrix
  File "/home/runner/work/rust/rust/src/ci/github-actions/ci.py", line 266, in calculate_job_matrix
    jobs = calculate_jobs(run_type, job_data)
  File "/home/runner/work/rust/rust/src/ci/github-actions/ci.py", line 153, in calculate_jobs
    raise Exception(
    raise Exception(
Exception: Custom job(s) `['x86_64-msvc']` not found in auto jobs
##[error]Process completed with exit code 1.
```

</details>

tgross35 committed:
Improve debugging for metadata structures

I had to do a lot of debug by printing; having these `Debug` traits in
place made it easier. Additionally, add some more information to
existing `info!` statements.

tgross35 committed:
Inject `compiler_builtins` during postprocessing rather than via AST

`compiler_builtins` is currently injected as `extern crate
compiler_builtins as _`. This has made gating via diagnostics difficult
because it appears in the crate graph as a non-private dependency, and
there isn't an easy way to differentiate between the injected AST and
user-specified `extern crate compiler_builtins`.

Resolve this by injecting `compiler_builtins` during postprocessing
rather than early in the AST. Most of the time this isn't even needed
because it shows up in `std` or `core`'s crate graph, but injection is
still needed to ensure `#![no_core]` works correctly.

A similar change was attempted at [1] but this encountered errors
building `proc_macro` and `rustc-std-workspace-std`. Similar failures
showed up while working on this patch, which were traced back to
`compiler_builtins` showing up in the graph twice (once via dependency
and once via injection). This is resolved by not injecting if a
`#![compiler_builtins]` crate already exists.

[1]: https://github.com/rust-lang/rust/pull/113634

tgross35 commented:
@bors try

bors commented:
:hourglass: Trying commit 081741bf9b43bbcf590d0ea6b759ed212af28625 with merge c649e3a1a74174e44994f7d653e0abc015457d26...
<!-- homu: {"type":"TryBuildStarted","head_sha":"081741bf9b43bbcf590d0ea6b759ed212af28625","merge_sha":"c649e3a1a74174e44994f7d653e0abc015457d26"} -->

tgross35 commented on tests/ui/proc-macro/issue-59191-replace-root-with-fn.stderr:
I don't really follow what the change is here; were inner attribute macros removing the `extern crate compiler_builtins` and now this error is showing up because `compiler_builtins` is somehow requiring a panic handler? That doesn't seem right.

In any case, this output change seems unproblematic.

tgross35 commented:
@rustbot ready (thanks @petrochenkov for taking a look)

rust-log-analyzer commented:
The job **`test-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13042899084/job/36388378241) [(plain)](https://github.com/rust-lang-ci/rust/commit/c649e3a1a74174e44994f7d653e0abc015457d26/checks/36388378241/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
diff of stderr:

8   --> $DIR/assoc-const-eq-missing.rs:16:16
9    |
10 LL | fn foo2<F: Foo<Z = usize>>() {}
-    |                ^ associated type `Z` not found
13 error[E0220]: associated constant `Z` not found for `Foo`
14   --> $DIR/assoc-const-eq-missing.rs:18:16



The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args associated-consts/assoc-const-eq-missing.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/associated-consts/assoc-const-eq-missing.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=wasm32-wasip1" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-consts/assoc-const-eq-missing" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/wasm32-wasip1/native/rust-test-helpers"
--- stderr -------------------------------
error[E0220]: associated constant `Z` not found for `Foo`
##[error]  --> /checkout/tests/ui/associated-consts/assoc-const-eq-missing.rs:14:16
   |
   |
LL | fn foo1<F: Foo<Z = 3>>() {}
   |                ^ help: there is an associated constant with a similar name: `N`
error[E0220]: associated type `Z` not found for `Foo`
##[error]  --> /checkout/tests/ui/associated-consts/assoc-const-eq-missing.rs:16:16
   |
   |
LL | fn foo2<F: Foo<Z = usize>>() {}
   |                ^ there is a similarly named associated type `H` in the trait `compiler_builtins::int::traits::DInt`
error[E0220]: associated constant `Z` not found for `Foo`
##[error]  --> /checkout/tests/ui/associated-consts/assoc-const-eq-missing.rs:18:16
   |
   |
LL | fn foo3<F: Foo<Z = 5>>() {}
   |                ^ help: there is an associated constant with a similar name: `N`
error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0220`.
------------------------------------------
---
diff of stderr:

2   --> $DIR/E0220.rs:5:22
3    |
4 LL | type Foo = dyn Trait<F=i32>;
-    |                      ^ help: `Trait` has the following associated type: `Bar`
7 error[E0191]: the value of the associated type `Bar` in `Trait` must be specified
8   --> $DIR/E0220.rs:5:16



The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args error-codes/E0220.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/error-codes/E0220.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=wasm32-wasip1" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/error-codes/E0220" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/wasm32-wasip1/native/rust-test-helpers"
--- stderr -------------------------------
error[E0220]: associated type `F` not found for `Trait`
##[error]  --> /checkout/tests/ui/error-codes/E0220.rs:5:22
   |
   |
LL | type Foo = dyn Trait<F=i32>; //~ ERROR E0220
   |                      ^ there is a similarly named associated type `H` in the trait `compiler_builtins::int::traits::DInt`
error[E0191]: the value of the associated type `Bar` in `Trait` must be specified
##[error]  --> /checkout/tests/ui/error-codes/E0220.rs:5:16
   |
LL |     type Bar;
LL |     type Bar;
   |     -------- `Bar` defined here
...
LL | type Foo = dyn Trait<F=i32>; //~ ERROR E0220
   |                ^^^^^^^^^^^^ help: specify the associated type: `Trait<F=i32, Bar = Type>`
error: aborting due to 2 previous errors

Some errors have detailed explanations: E0191, E0220.
For more information about an error, try `rustc --explain E0191`.
---
diff of stderr:

54   --> $DIR/issue-95023.rs:8:44
55    |
56 LL |     fn foo<const N: usize>(&self) -> Self::B<{ N }>;
-    |                                            ^ help: `Self` has the following associated type: `Output`
59 error: aborting due to 7 previous errors
60 



The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lifetimes/issue-95023.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lifetimes/issue-95023.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=wasm32-wasip1" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lifetimes/issue-95023" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/wasm32-wasip1/native/rust-test-helpers"
--- stderr -------------------------------
error: associated function in `impl` without body
##[error]  --> /checkout/tests/ui/lifetimes/issue-95023.rs:8:5
   |
   |
LL |     fn foo<const N: usize>(&self) -> Self::B<{ N }>;
   |                                                    |
   |                                                    |
   |                                                    help: provide a definition for the function: `{ <body> }`
error[E0407]: method `foo` is not a member of trait `Fn`
##[error]  --> /checkout/tests/ui/lifetimes/issue-95023.rs:8:5
   |
   |
LL |     fn foo<const N: usize>(&self) -> Self::B<{ N }>;

error[E0183]: manual implementations of `Fn` are experimental
##[error]  --> /checkout/tests/ui/lifetimes/issue-95023.rs:3:6
   |
---
   |
LL | impl Fn(&isize) for Error {
   |      ^^^^^^^^^^ associated item constraint not allowed here
   |
help: parenthesized trait syntax expands to `Fn<(&isize,), Output=()>`
   |
LL | impl Fn(&isize) for Error {
   |      ^^^^^^^^^^


error[E0277]: expected a `FnMut(&isize)` closure, found `Error`
   |
LL | impl Fn(&isize) for Error {
LL | impl Fn(&isize) for Error {
   |                     ^^^^^ expected an `FnMut(&isize)` closure, found `Error`
   |
   = help: the trait `FnMut(&isize)` is not implemented for `Error`
  --> /rustc/FAKE_PREFIX/library/core/src/ops/function.rs:76:1

error[E0046]: not all trait items implemented, missing: `call`
##[error]  --> /checkout/tests/ui/lifetimes/issue-95023.rs:3:1
##[error]  --> /checkout/tests/ui/lifetimes/issue-95023.rs:3:1
   |
LL | impl Fn(&isize) for Error {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^ missing `call` in implementation
   |
   = help: implement the missing item: `fn call(&self, _: (&isize,)) -> <Self as FnOnce<(&isize,)>>::Output { todo!() }`
error[E0220]: associated type `B` not found for `Self`
##[error]  --> /checkout/tests/ui/lifetimes/issue-95023.rs:8:44
   |
   |
LL |     fn foo<const N: usize>(&self) -> Self::B<{ N }>;
   |                                            ^ there is a similarly named associated type `H` in the trait `compiler_builtins::int::traits::DInt`
error: aborting due to 7 previous errors

Some errors have detailed explanations: E0046, E0183, E0220, E0229, E0277, E0407.
For more information about an error, try `rustc --explain E0046`.
---
1 error: `#[panic_handler]` function required, but not found
2 
- error: unwinding panics are not supported without std
-    |
-    = help: using nightly cargo, use -Zbuild-std with panic="abort" to avoid unwinding
-    = note: since the core library is usually precompiled with panic="unwind", rebuilding your crate with panic="abort" may not be enough to fix the problem
8 error: requires `sized` lang_item
9 

- error: aborting due to 3 previous errors
---
To only update this specific test, also pass `--test-args proc-macro/issue-59191-replace-root-with-fn.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/proc-macro/issue-59191-replace-root-with-fn.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=wasm32-wasip1" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/issue-59191-replace-root-with-fn" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/wasm32-wasip1/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/issue-59191-replace-root-with-fn/auxiliary" "--edition=2018" "--extern" "issue_59191=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/issue-59191-replace-root-with-fn/auxiliary/libissue_59191.so"
--- stderr -------------------------------
error: `#[panic_handler]` function required, but not found

error: requires `sized` lang_item
```

</details>

tgross35 commented:
The above failure came from trying to do the compiler-builtins update here too, but the pub/private fixes from https://github.com/rust-lang/rust/pull/135501 are still needed to make that work. To keep the changes smaller I'll leave this PR as only the injection change, then rebase https://github.com/rust-lang/rust/pull/135501 on top to fix the rest and do the update.

@bors try

bors commented:
:hourglass: Trying commit 4bc6910b18850d8b2e63773cf87f5b70c3646b49 with merge 34777d9575856ebf3dddbb7db0cbe5a6d27442d5...
<!-- homu: {"type":"TryBuildStarted","head_sha":"4bc6910b18850d8b2e63773cf87f5b70c3646b49","merge_sha":"34777d9575856ebf3dddbb7db0cbe5a6d27442d5"} -->

rust-log-analyzer commented:
The job **`test-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13044055819/job/36391529706) [(plain)](https://github.com/rust-lang-ci/rust/commit/34777d9575856ebf3dddbb7db0cbe5a6d27442d5/checks/36391529706/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
1 error: `#[panic_handler]` function required, but not found
2 
- error: unwinding panics are not supported without std
-    |
-    = help: using nightly cargo, use -Zbuild-std with panic="abort" to avoid unwinding
-    = note: since the core library is usually precompiled with panic="unwind", rebuilding your crate with panic="abort" may not be enough to fix the problem
8 error: requires `sized` lang_item
9 

- error: aborting due to 3 previous errors
---
To only update this specific test, also pass `--test-args proc-macro/issue-59191-replace-root-with-fn.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/proc-macro/issue-59191-replace-root-with-fn.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=wasm32-wasip1" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/issue-59191-replace-root-with-fn" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/wasm32-wasip1/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/issue-59191-replace-root-with-fn/auxiliary" "--edition=2018" "--extern" "issue_59191=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/issue-59191-replace-root-with-fn/auxiliary/libissue_59191.so"
--- stderr -------------------------------
error: `#[panic_handler]` function required, but not found

error: requires `sized` lang_item
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13044055819/job/36392976288)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13044055819/job/36392976288","builder_name":"checks-actions"} -->

tgross35 commented:
@bors try

bors commented:
:hourglass: Trying commit 55facbb2a88d36875b93a386c1a11ba3ffa63338 with merge a3ee37797750823864ab6d0bccb25f67c60a5bb0...
<!-- homu: {"type":"TryBuildStarted","head_sha":"55facbb2a88d36875b93a386c1a11ba3ffa63338","merge_sha":"a3ee37797750823864ab6d0bccb25f67c60a5bb0"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13045091884/job/36394251240) [(plain)](https://github.com/rust-lang-ci/rust/commit/a3ee37797750823864ab6d0bccb25f67c60a5bb0/checks/36394251240/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  98% (50635/51668)
Updating files:  99% (51152/51668)
Updating files: 100% (51668/51668)
Updating files: 100% (51668/51668), done.
branch 'try' set up to track 'origin/try'.
Switched to a new branch 'try'
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format=%H
a3ee37797750823864ab6d0bccb25f67c60a5bb0
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
---- [run-make] tests\run-make\arguments-non-c-like-enum stdout ----

error: rmake recipe failed to complete
status: exit code: 1
command: "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\run-make\\arguments-non-c-like-enum\\rmake.exe"
--- stderr -------------------------------
--- stderr -------------------------------
"C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.42.34433\\bin\\HostX64\\x64\\cl.exe" "-nologo" "-MT" "-Brepro" "test.c" "nonclike.lib" "-Fe:test.exe" "-Fo:test.obj" "ws2_32.lib" "userenv.lib" "advapi32.lib" "bcrypt.lib" "ntdll.lib" "synchronization.lib"
command failed at line 14
Command { cmd: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.42.34433\\bin\\HostX64\\x64\\cl.exe" "-nologo" "-MT" "-Brepro" "test.c" "nonclike.lib" "-Fe:test.exe" "-Fo:test.obj" "ws2_32.lib" "userenv.lib" "advapi32.lib" "bcrypt.lib" "ntdll.lib" "synchronization.lib", stdin_buf: None, stdin: None, stdout: None, stderr: None, drop_bomb: DropBomb { command: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.42.34433\\bin\\HostX64\\x64\\cl.exe", defused: true, armed_location: Location { file: "D:\\a\\rust\\rust\\tests\\run-make\\arguments-non-c-like-enum\\rmake.rs", line: 8, col: 5 } }, already_executed: true }
output status: `exit code: 2`
=== STDOUT ===
test.c
nonclike.lib(core-2d52bca1968b96fb.core.da0242bda8ac3051-cgu.0.rcgu.o) : error LNK2019: unresolved external symbol __udivti3 referenced in function _ZN4core3fmt3num8exp_u12817h2d9af119bb1953d9E



=== STDERR ===
=== STDERR ===
------------------------------------------


---- [run-make] tests\run-make\c-link-to-rust-va-list-fn stdout ----

error: rmake recipe failed to complete
status: exit code: 1
command: "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\run-make\\c-link-to-rust-va-list-fn\\rmake.exe"
--- stderr -------------------------------
command failed at line 16
command failed at line 16
Command { cmd: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.42.34433\\bin\\HostX64\\x64\\cl.exe" "-nologo" "-MT" "-Brepro" "test.c" "checkrust.lib" "-Fe:test.exe" "-Fo:test.obj" "ws2_32.lib" "userenv.lib" "advapi32.lib" "bcrypt.lib" "ntdll.lib" "synchronization.lib", stdin_buf: None, stdin: None, stdout: None, stderr: None, drop_bomb: DropBomb { command: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.42.34433\\bin\\HostX64\\x64\\cl.exe", defused: true, armed_location: Location { file: "D:\\a\\rust\\rust\\tests\\run-make\\c-link-to-rust-va-list-fn\\rmake.rs", line: 12, col: 5 } }, already_executed: true }
output status: `exit code: 2`
=== STDOUT ===
test.c
checkrust.lib(core-2d52bca1968b96fb.core.da0242bda8ac3051-cgu.0.rcgu.o) : error LNK2019: unresolved external symbol __udivti3 referenced in function _ZN4core3fmt3num8exp_u12817h2d9af119bb1953d9E



=== STDERR ===
=== STDERR ===
------------------------------------------


---- [run-make] tests\run-make\lto-smoke-c stdout ----

error: rmake recipe failed to complete
status: exit code: 1
command: "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\run-make\\lto-smoke-c\\rmake.exe"
--- stderr -------------------------------
command failed at line 18
command failed at line 18
Command { cmd: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.42.34433\\bin\\HostX64\\x64\\cl.exe" "-nologo" "-MT" "-Brepro" "bar.c" "foo.lib" "-Fe:bar.exe" "-Fo:bar.obj" "ws2_32.lib" "userenv.lib" "advapi32.lib" "bcrypt.lib" "ntdll.lib" "synchronization.lib", stdin_buf: None, stdin: None, stdout: None, stderr: None, drop_bomb: DropBomb { command: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.42.34433\\bin\\HostX64\\x64\\cl.exe", defused: true, armed_location: Location { file: "D:\\a\\rust\\rust\\tests\\run-make\\lto-smoke-c\\rmake.rs", line: 13, col: 5 } }, already_executed: true }
output status: `exit code: 2`
=== STDOUT ===
bar.c
foo.lib(foo.foo.5be5606e1f6aa79b-cgu.0.rcgu.o) : error LNK2019: unresolved external symbol __udivti3 referenced in function _ZN4core3fmt3num8exp_u12817h2d9af119bb1953d9E
bar.exe : fatal error LNK1120: 1 unresolved externals


=== STDERR ===
------------------------------------------
------------------------------------------


---- [run-make] tests\run-make\msvc-wholearchive stdout ----

error: rmake recipe failed to complete
status: exit code: 1
command: "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\run-make\\msvc-wholearchive\\rmake.exe"
--- stderr -------------------------------
command failed at line 35
command failed at line 35
Command { cmd: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.42.34433\\bin\\HostX64\\x64\\link.exe" "c.obj" "./static.lib" "-dll" "-def:dll.def" "-out:dll.dll" "ws2_32.lib" "userenv.lib" "advapi32.lib" "bcrypt.lib" "ntdll.lib" "synchronization.lib", stdin_buf: None, stdin: None, stdout: None, stderr: None, drop_bomb: DropBomb { command: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.42.34433\\bin\\HostX64\\x64\\link.exe", defused: true, armed_location: Location { file: "D:\\a\\rust\\rust\\tests\\run-make\\msvc-wholearchive\\rmake.rs", line: 32, col: 5 } }, already_executed: true }
output status: `exit code: 1120`
=== STDOUT ===
Microsoft (R) Incremental Linker Version 14.42.34436.0


   Creating library dll.lib and object dll.exp
static.lib(core-2d52bca1968b96fb.core.da0242bda8ac3051-cgu.0.rcgu.o) : error LNK2019: unresolved external symbol __udivti3 referenced in function _ZN4core3fmt3num8exp_u12817h2d9af119bb1953d9E
dll.dll : fatal error LNK1120: 1 unresolved externals


=== STDERR ===
------------------------------------------
------------------------------------------


---- [run-make] tests\run-make\reachable-extern-fn-available-lto stdout ----

error: rmake recipe failed to complete
status: exit code: 1
command: "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\run-make\\reachable-extern-fn-available-lto\\rmake.exe"
--- stderr -------------------------------
command failed at line 24
command failed at line 24
Command { cmd: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.42.34433\\bin\\HostX64\\x64\\cl.exe" "-nologo" "-MT" "-Brepro" "foo.c" "bar.lib" "ws2_32.lib" "userenv.lib" "advapi32.lib" "bcrypt.lib" "ntdll.lib" "synchronization.lib" "-Fe:foo.exe" "-Fo:foo.obj", stdin_buf: None, stdin: None, stdout: None, stderr: None, drop_bomb: DropBomb { command: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.42.34433\\bin\\HostX64\\x64\\cl.exe", defused: true, armed_location: Location { file: "D:\\a\\rust\\rust\\tests\\run-make\\reachable-extern-fn-available-lto\\rmake.rs", line: 24, col: 5 } }, already_executed: true }
output status: `exit code: 2`
=== STDOUT ===
foo.c
bar.lib(bar.bar.67338c9292c4797c-cgu.0.rcgu.o) : error LNK2019: unresolved external symbol __udivti3 referenced in function _ZN4core3fmt3num8exp_u12817h2d9af119bb1953d9E
foo.exe : fatal error LNK1120: 1 unresolved externals


=== STDERR ===
------------------------------------------
---
test result: FAILED. 280 passed; 5 failed; 90 ignored; 0 measured; 6 filtered out; finished in 203.58s

Some tests failed in compiletest suite=run-make mode=run-make host=x86_64-pc-windows-msvc target=x86_64-pc-windows-msvc
Build completed unsuccessfully in 2:24:59
make: *** [Makefile:109: ci-msvc-py] Error 1
  network time: Thu, 30 Jan 2025 05:58:41 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13045091884/job/36397921693)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13045091884/job/36397921693","builder_name":"checks-actions"} -->

tgross35 commented:
...so close.

All of the failed tests build `staticlib`s, postprocessing must not run with that option.

tgross35 commented:
It looks like `Implicit` means it doesn't get linked in static libs https://github.com/rust-lang/rust/blob/e6f12c8b7d8d5c821c32fb2739ea01d1d874c58a/compiler/rustc_metadata/src/dependency_format.rs#L350-L362. Trying again with implicit -> explicit.

@bors try

bors commented:
:hourglass: Trying commit 01124991d6002f899f2988c8e2bb550f9e7f6819 with merge 9bd0b76dacf5bdb1e1667d010cfd6e26268dcc36...
<!-- homu: {"type":"TryBuildStarted","head_sha":"01124991d6002f899f2988c8e2bb550f9e7f6819","merge_sha":"9bd0b76dacf5bdb1e1667d010cfd6e26268dcc36"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13048569691/job/36411323542)
Build commit: 9bd0b76dacf5bdb1e1667d010cfd6e26268dcc36 (`9bd0b76dacf5bdb1e1667d010cfd6e26268dcc36`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13048569691/job/36411323542"},"merge_sha":"9bd0b76dacf5bdb1e1667d010cfd6e26268dcc36"} -->

tgross35 commented:
https://github.com/rust-lang/rust/pull/135501 includes this PR plus the couple of remaining changes to make the update actually work. Since the difference between the two is reasonably small, I am going to close this PR in favor of that one (fyi @petrochenkov if you were planning to review).
